src/formats/MOL2.cpp

```diff
@@ -0,0 +1,298 @@
+  : file_(TextFile::create(path, mode)) {
+    while (!file_->eof()) {
+        auto position = forward(*file_);
+        if (position == std::streampos(-1)) {
+            break;
+        }
+        if (!file_) {
+            throw format_error("IO error while reading '{}' as MOL2", path);
+        }
+        steps_positions_.push_back(position);
+    }
+    file_->rewind();
+}
+
+size_t MOL2Format::nsteps() {
+    return steps_positions_.size();
+}
+
+void MOL2Format::read_step(const size_t step, Frame& frame) {
+    assert(step < steps_positions_.size());
+    file_->seekg(steps_positions_[step]);
+    read(frame);
+}
+
+void MOL2Format::read(Frame& frame) {
+    const auto& startline= file_->readline();
+    assert(startline.substr(0, 17) == "@<TRIPOS>MOLECULE");
+
+    frame.set("name", trim(file_->readline()) );
+    const auto& line = file_->readline();
+
+    const auto counts = split(line, ' ');
+
+    long long natoms = string2longlong(counts[0]);
+    long long nbonds = 0;
+    
+    if (counts.size() >= 2) {
+        nbonds = string2longlong(counts[1]);
+    }
+
+    residues_.clear();
+    frame.resize(0);
+    frame.reserve(static_cast<size_t>(natoms));
+
+    auto junk = file_->readline();
+    bool charges = (trim(file_->readline()) != "NO_CHARGES");
+
+    while (!file_->eof()) {
+        auto curr_pos = file_->tellg();
+        auto line2 = trim(file_->readline());
+
+        if (line2 == "@<TRIPOS>ATOM") {
+            read_atoms(frame, static_cast<size_t>(natoms), charges);
+        } else if (line2 == "@<TRIPOS>BOND") {
+            read_bonds(frame, static_cast<size_t>(nbonds));
+        } else if (line2 == "@<TRIPOS>CRYSIN") {
+            auto cryst = file_->readline();
+
+            double a, b, c, alpha, beta, gamma;
+            scan(cryst, " %lf %lf %lf %lf %lf %lf",
+                &a, &b, &c, &alpha, &beta, &gamma
+            );
+
+            frame.set_cell(UnitCell(a, b, c, alpha, beta, gamma));
+        } else if (line2 == "@<TRIPOS>MOLECULE") {
+            file_->seekg(curr_pos);
+            break;
+        }
+    }
+
+    for (auto& residue: residues_) {
+        frame.add_residue(residue.second);
+    }
+
+}
+
+void MOL2Format::read_atoms(Frame& frame, size_t natoms, bool charges) {
+    auto lines = file_->readlines(natoms);
+
+    for (const auto& line : lines) {
+        unsigned long id, resid;
+        char atom_name[32], atom_type[32], res_name[32];
+        double x, y, z;
+        double charge = 0;
+
+        if (charges) {
+            scan(line, " %lu %31s %lf %lf %lf %31s %lu %31s %lf",
+                &id, &atom_name[0], &x, &y, &z, &atom_type[0], &resid, &res_name[0], &charge
+            );
+        } else {
+            scan(line, " %lu %31s %lf %lf %lf %31s %lu %31s",
+                &id, &atom_name[0], &x, &y, &z, &atom_type[0], &resid, &res_name[0]
+            );
+        }
+
+        Atom atom(atom_name);
+        frame.add_atom(std::move(atom), Vector3D(x, y, z));
```

Comment:
Why not creating the atom directly here, instead of creating it and then moving it?

You can also use initializer lists to create the vector with less noise
```cpp
frame.add_atom(Atom(atom_name, atom_type), {x, y, z});
```